src/simulators/statevector/qubitvector.hpp

```diff
@@ -516,128 +684,120 @@ cvector_t QubitVector<statevector_t>::vector() const {
   if (!checkpoint_)
-    checkpoint_ = reinterpret_cast<complex_t*>(malloc(sizeof(complex_t) * num_states_));
+    checkpoint_ = reinterpret_cast<complex_t*>(malloc(sizeof(complex_t) * size_));
 
-  const int_t end = num_states_;    // end for k loop
+  const int_t end = size_;    // end for k loop
 #pragma omp parallel for if (num_qubits_ > omp_threshold_ && omp_threads_ > 1) num_threads(omp_threads_)
   for (int_t k = 0; k < end; ++k)
-    checkpoint_[k] = statevector_[k];
+    checkpoint_[k] = data_[k];
 }
 
-template <class statevector_t>
-complex_t QubitVector<statevector_t>::inner_product() const {
+
+template <class data_t>
+void QubitVector<data_t>::revert(bool keep) {
 
   #ifdef DEBUG
   check_checkpoint();
   #endif
 
-  double z_re = 0., z_im = 0.;
-  const int_t end = num_states_;    // end for k loop
-#pragma omp parallel reduction(+:z_re, z_im) if (num_qubits_ > omp_threshold_ && omp_threads_ > 1) num_threads(omp_threads_)
-  {
-  #pragma omp for
-    for (int_t k = 0; k < end; k++) {
-      const complex_t z = statevector_[k] * std::conj(checkpoint_[k]);
-      z_re += std::real(z);
-      z_im += std::imag(z);
-    }
-  } // end omp parallel
-  return complex_t(z_re, z_im);
+  const int_t end = size_;    // end for k loop
+#pragma omp parallel for if (num_qubits_ > omp_threshold_ && omp_threads_ > 1) num_threads(omp_threads_)
+  for (int_t k = 0; k < end; ++k)
+    data_[k] = checkpoint_[k];
+
+  if (!keep) {
+    free(checkpoint_);
+    checkpoint_ = 0;
+  }
 }
 
-template <class statevector_t>
-void QubitVector<statevector_t>::revert(bool keep) {
+template <class data_t>
+complex_t QubitVector<data_t>::inner_product() const {
 
   #ifdef DEBUG
   check_checkpoint();
   #endif
+  // Lambda function for inner product with checkpoint state
+  auto lambda = [&](int_t k, double &val_re, double &val_im)->void {
+    const complex_t z = data_[k] * std::conj(checkpoint_[k]);
+    val_re += std::real(z);
+    val_im += std::imag(z);
+  };
+  return apply_reduction_lambda(lambda);
+}
+
+//------------------------------------------------------------------------------
+// Initialization
+//------------------------------------------------------------------------------
+
+template <class data_t>
+void QubitVector<data_t>::initialize() {
+  zero();
+  data_[0] = 1.;
+}
+
+template <class data_t>
+void QubitVector<data_t>::initialize_from_vector(const cvector_t &statevec) {
+  if (size_ != statevec.size()) {
+    std::stringstream ss;
+    ss << "QubitVector::initialize input vector is incorrect length (";
+    ss << size_ << "!=" << statevec.size() << ")";
+    throw std::runtime_error(ss.str());
+  }
+
+  const int_t end = size_;    // end for k loop
 
-  const int_t end = num_states_;    // end for k loop
 #pragma omp parallel for if (num_qubits_ > omp_threshold_ && omp_threads_ > 1) num_threads(omp_threads_)
   for (int_t k = 0; k < end; ++k)
-    statevector_[k] = checkpoint_[k];
+    data_[k] = statevec[k];
```

Comment:
`std::copy()` ?

Reply:
I didn't std::copy work on raw c-arrays? These functions were coded this way to be agnostic to the data type (as long as it has an accessor []).